### PR TITLE
Bindings redirects fix

### LINKED ISSUE #865: Error on dotnet run
Guys,

I'm getting an error when running dotnet run or dotnet build:
_An item with the same key has already been added. Key: (log4net, neutral, 669e0ddf0bb1aa2a)_

Not sure how to proceed.

Thank you!

A bit more context:
My project.lock.json contains the following bits:

```
"Log4Net/2.0.5": {
        "type": "package",
        "compile": {
          "lib/net45-full/log4net.dll": {}
        },
        "runtime": {
          "lib/net45-full/log4net.dll": {}
        }
      },
      "log4net.Ext.Json.1.2.13/1.2.13.20140804": {
        "type": "package",
        "dependencies": {
        "log4net": "2.0.3"
        },
        "compile": {
          "lib/net35-full/log4net.Ext.Json.dll": {}
        },
        "runtime": {
          "lib/net35-full/log4net.Ext.Json.dll": {}
        }
      }
```

Notice the log4net dependency of log4net.Ext.Json has different casing than the root Log4Net dependency. Could this be the issue?

UPDATE2:
Changed the cached nuspec file for log4net.Ext.Json to say Log4Net and it made the error go away. At this point I think this is a bug. I'll try and have a look at the code.

### PR DESCRIPTION
**Pull request by marono**
Due to packages including assemblies other than their primary output,
under some circumstances, the same assembly may be considered multiple
times for bindign redirect generation, which would cause the build to
fail because of dictionary key collisions. Fixed the issue by filtering
the assemblies considered for binding redirects to only include
exports in the runtime list only when the same export is not already
published via a compilation list of another dependency.

Fix #865

### CODE REVIEW

**Comment by marono** [comment]
How do I prevent my test from running on non-Windows platforms? Added WindowsOnlyFactAttribute, but doesn't seem to do the trick.

**Comment by piotrpMSFT** [comment]
@marono this is blocked on both CI and CLA.

**Comment by marono** [comment]
@piotrpMSFT how do I prevent my tests from running on OSX/Linux? Because binding redirects are only generated for 'Desktop' runtimes, all my test projects are targeted for net461.

Not sure why the Debug build on windows fails, I'll investigate it as soon as the CI server comes back up.

**Comment by marono** [comment]
@piotrpMSFT, windows debug build failed because of a timeout. Don't think that's related to my changes, could you trigger it again please?

File: src/Microsoft.DotNet.Compiler.Common/BindingRedirectGenerator.cs [hunk 1]
```diff
@@ -109,26 +109,24 @@ private static XElement GetOrAddElement(XContainer parent, XName elementName)
 
         private static AssemblyRedirect[] CollectRedirects(IEnumerable<LibraryExport> dependencies)
         {
-            var allRuntimeAssemblies = dependencies.SelectMany(d => d.RuntimeAssemblies).Select(GetAssemblyInfo).ToArray();
-            var assemblyLookup = allRuntimeAssemblies.ToDictionary(r => r.Identity.ToLookupKey());
+            var dependenciesList = dependencies.ToList();
+            var allRuntimeAssemblies = dependenciesList.SelectMany(d => d.RuntimeAssemblies).Select(GetAssemblyInfo).ToArray();
+            var assemblyLookup = GetListOfUniqueRefs(dependenciesList).Select(GetAssemblyInfo).ToDictionary(r => r.Identity.ToLookupKey());
```
Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/BindingRedirectGenerator.cs]: You are calling GetAssemblyInfo on some assemblies twice, I'd prefer not doing this.

**Comment by marono** [comment]
Now I don't understand it, yesterday Windows Release was green, but not anymore. What's changed?

**Comment by marono** [comment]
Guys,

This is ready for merging, I've just signed my CLA, sorry for the delay, it's the first time my employer does this so they had to check the agreement.

I've also amended the code to cache assembly info during scanning.

For some reason, builds on non-Windows machines are not triggered, maybe you know why that is?

**Comment by marono** [comment]
I've been asked to resubmit the PR as there was an issue with my CLA form, I'll close this one for now.

**Pull request closed without merging**